### ISSUE TITLE
How should these definitions be consumed?

### ISSUE THREAD

**Issue by splodingsocks** [flowtype/flow-typed#4]
As mentioned by @danvk, we need to figure out how best to instruct others to obtain and use these definitions. I'll be back next week to discuss more.

**Comment by jeffmo** [comment]
cc @gabelevi who's working on some related things WRT interface definitions that can easily be shipped alongside (but decoupled from) the library implementation

**Comment by danvk** [comment]
I do like the idea of libraries distributing their own typings.

Some pros:
- typings are tied to the version of the library to which they apply
- Installation is simple (type definitions come with the node module)

Some cons:
- Owners of major libraries may not care to distribute type definitions (especially older ones which aren't updated often)
- When there's a new version of Flow, it's harder for the Flow owners to check for regressions against _all_ type definitions (which they could do with this repo).

**Comment by jeffmo** [comment]
> When there's a new version of Flow, it's harder for the Flow owners to check for regressions against all type definitions (which they could do with this repo).

Yea agree with the problem, though I see this as a separate problem that we need to work on even independent of interface definitions.

One thing that we do with our codebases to address a part of this is to explicitly set the `version=` option in the config. Then we use a wrapper script for the `flow` command that reads this and then runs the appropriate version of flow (rather than a single global version of Flow installed). I really think this is the right direction for the command you should get through homebrew/opam/flow-cli.

It doesn't quite solve another problem, though: What to do when two libraries used by a single project dictate that they rely on disjoint versions of flow? For this I don't have a good answer yet...but it's probably something we need to consider soon. We haven't had to address it with our own projects yet because we don't sub-version project dependencies...but with npm and such there's no getting around it as a common case.

**Comment by magicpat** [comment]
@jeffmo As a starter, since there will be a lot of unknowns how to do the versioning and similar stuff... would it be a good idea to start a new cli-tool to retrieve definitions from this repository, similar how DefinitelyTyped's cli-tool does? I would be keen to invest some time in this and get something going.

**Comment by splodingsocks** [comment]
@magicpat I think it's for sure a good idea. In fact, I think @jeffmo has already done some thinking in this area.

**Comment by hallettj** [comment]
In my opinion, the simplest useful option would be to push FlowTyped to npm as-is. Users could include it as a dependency, and add lines in `.flowconfig` to point to the type definitions that they want to use. For example:

```
[libs]
node_modules/FlowTyped/underscore/underscore.js
```

That would make it hard for consumers to pick and choose different versions of type definition files. But it would get something useful out into the world quickly.

A simple way to make multiple versions of definitions available for a single file would be to encode versions in file paths. Eg, `node_modules/FlowTyped/underscore/1.8/underscore.js` vs `node_modules/FlowTyped/underscore/1.7/underscore.js`.

**Comment by splodingsocks** [comment]
Great idea, @hallettj. I've published it here:

https://www.npmjs.com/package/flow-typed

And I'll get to reviewing your PRs soon, thanks so much for contributing!

**Comment by jeffmo** [comment]
So I haven't had time to write all of this up yet, but I was actually hoping that we could use the npm package to do a little more of an interesting integration (you add it as a dependency and it automatically scrapes your package.json's list of dependencies and fetches the relevant libdefs for you).

I'm really hoping I can write these thoughts out in a more coherent way this week so we can talk through them, but I'd really like for this repo to be a defacto part of using Flow with npm

**Comment by jeffmo** [comment]
One downside to publishing the entire repo is just that it means you'll get libdefs for packages that you don't actually have installed. I think it'll be important to keep the installed libdefs as close to reality as possible. There are also versioning concerns to be taken carefully into account.

**Comment by splodingsocks** [comment]
Right, good points, @jeffmo. It shouldn't be a problem to have published what we have now, and change it when we get the rest figured out, should it?

**Comment by jeffmo** [comment]
Oh no I don't see any issues here -- mostly just wanted to lead in with some context :)

**Comment by magicpat** [comment]
I don't want to spoil this idea with premature optimization issues, but if I look at DefinitelyTyped, I would rather not have to install thousands of files (especially for every project I install flow-typed in my devDependencies). 

But I really like the versioning idea i.e. `node_modules/FlowTyped/underscore/1.0/underscore.js`.
If we concise to a proper versioning structure and maybe add something like a `underscore/1.0/flow_version` file for describing compatible flow versions I will additionally create an complementary commandline tool for those who do not want to pull the whole repo.

Example of `underscore/1.0/flow_version`:

```
{
  "flow": "~0.22"
}
```

## CLI concept

Just so I have this somewhere in place and for the sake of this discussion, I will write this here...

**Interface of the cli-tool:**

```sh
# Reads ./flowtyped.json for configuration
flow-typed install <dependency>
```
**Config file `<Project>/flowtyped.json`**:

```javascript
//Default-Values
{
   "source": "https://github.com/splodingsocks/FlowTyped",
   "destination": "interfaces/"

   //Will parse the package.json file for dependency-versions and
   //then tries do match a proper declaration file to it from our source
   "autodetect": true,

   "dependencies":  {
      //i.e. I want this exact version
      "underscore": "1.0",

      //or maybe enable me to use my own local files (which will be copied / symlinked)
      "immutablejs": "~/flow_declarations/own_immutable.js"
   } 
}
```

The tool could also check if `.flowconfig` has an `[included]` configuration with proper `destination` set, otherwise it may put out a warning or autofixes it.

**Comment by jeffmo** [comment]
@magicpat: Yea, I agree -- I think we can do a lot better than DefinitelyTyped in these areas.

Between good versioning, automating libdef generation where possible (I'll go into this in more detail a little later), a good libdef testing setup, and making it super easy get exactly the libdefs you need at the version you need with as little effort as possible, I think we can MakeLibdefsGreatAgain™.

As for a CLI tool -- this looks great and I think we might need one at some point, but I actually think we should explore options that require minimal configuration (if any?).

For example: I was thinking it would be cool if the flow-typed npm package just had a postinstall script that actually fetched the correct lib-version/flow-version tuple from FlowTyped for you (based on your project's stated package.json dependencies) and placed that in a specific dir (say, `flow-typed?). Then Flow just knows to check that specific dir for libdefs out of the box.

There might still be a use for a CLI down the road that lets you do more specific things (I.e. test an update to a libdef, fetch specific libdefs ad-hoc, etc), but I'm betting the postinstall stuff could cover most of what we need at the outset.

**Comment by marudor** [comment]
Do we need a seperate config and CLI?
How about we just use npm?
Just found out about flow-typed - I'm managing some flow interfaces myself @ https://github.com/marudor/flowInterfaces
It started as a simple Repo I used as submodule. But it grew and some other people wanted to use it so I made a monorepo and just publish each libdef to npm (as iflow-<packageName>)
That way someone can just install it via NPM.
My only bummer is that I have to add each module to the [lib] section of flow manually.

Any thoughts?

**Comment by jeffmo** [comment]
@marudor: Yea, I agree that a new CLI/config shouldn't be the default/common-case usage.

Having individual npm packages is definitely an option, but the only annoying bit about this I can foresee is just that you have to explicitly (1) look to see if a lib def exists for a given dependency (2) specify both the real dependency and the libdef in your package.json (once for each dependency) (3) bump the versions of both the libdef and the real dep at the same time in the future.

Also @mroch brought up a good point yesterday that we should probably always install the lowest-compatible-version of a library definition for a given dependency (to ensure that Flow understands the most stable version of the libdef you have installed).

So for example:

Say that package 'foo' depends on package 'bar' like this:

```js
{
  "name": "foo",
  "version": "1.0.0",
  "dependencies": {
    "bar": "1.x"
  }
}
```

Package 'bar' has two versions, with interfaces like this:

```js
 // 1.0.0
declare function bar(x: string);

// 1.1.0
declare function bar(x: string);
declare function baz(y: number);
```

Per npm's semver rules, it's fine to use either version of 'bar' because 1.1.0 is backwards compatible with 1.0.0; its changes are purely additive.

### The problem

If I run npm install and get bar-1.1.0, and then add a call to baz(123), everything typechecks and I land it. But someone else, who `npm install`-ed before me and has bar-1.0.0, will get errors. Booooo, Flow failed us.

So to solve this, we should just always install the lowest-compatible-version of a libdef and, if someone wishes to use an API introduced in a higher/compatible version (say, bar-1.1.0) they just need to explicitly state that minimum-viable version in their package.json. Really they should do this anyway -- it was just easier to miss prior to tools like Flow warning them.

**Comment by jeffmo** [comment]
# Proposal

Ok, so to kick off a solid plan, here's a holistic proposal I'd like to put forth for discussion. Obviously we'll need to build this iteratively, but I'm willing to dedicate time to help build this (and would certainly be willing to work with anyone else who is interested in doing the same).

Also, this is totally a community discussion -- so if you see issues or things this doesn't address, let's talk through them and make sure we build something excellent here!

### FlowTyped Repo

This repo is the structured data-store for library definitions. Adding a new library definition is as simple as sending a pull request that includes the libdef, a specification of the package-version the libdef was generated for, and the version(s) of Flow the libdef is known to be compatible with.

A proposed file/directory structure for the repo is as follows:

* (dir) lodash `// exact npm package name`
  * (dir) v4 `// "v" + semver-comparable package version. "v4" matches *all* of "v4.x.x"`
    * (file) test-lodash-v4.1.js `// Lodash v4 test file. Runs for all versions of Flow`
    * (file) test-lodash-v4.2.js `// A second lodash v4 test file`
    * (dir) flow-all `// "-all" means "works with all Flow versions"`
      * (file) lodash-v4.js `// The flow-0.22 libdef. Always named "PKG-vVERSION.js"`
      * (file) test-flow-v0.22.js `// Flow 0.22-specific test. Hopefully these are rare`
      * (file) .flowconfig `// used for specifying the root of the test flow-project`
  * (dir) v3.2.1 `// Matches exactly "lodash@3.2.1". Most specific`
    * (file) test-lodash-v3.2.1.js
    * (dir) flow-v0.22 `// semver-comparable Flow version. More specific than "-all"`
      * (file) lodash-v3.2.1.js
      * (file) .flowconfig
* (dir) underscore
  * (dir) v1.8 `// Matches all of "v1.8.x"
    * (file) test-underscore.js
    * (dir) flow-all `// I expect "flow-all" to be most common`
      * (file) underscsore-v1.8.js
      * (file) .flowconfig
* ...

Directory structure conventions (which are checked/verified and relied on by tooling like Travis, an eventual `flow-typed` package, etc):

* Package dir names (toplevel dirs) must always match the npm package name exactly.
* Package-version sub-directories are always named `"v" + SEMVER_VERSION`.
* Flow-version sub-sub-directories are always named `"flow-v"  + FLOW_VERSION` *or* `"flow-all"`.
* There is intentionally no form of "-all" for package-version subdirectories. Each package-version dir *must* correspond to a published npm package version (or a semver string that resolves to at least one published npm package version).
* Test files must always be prefixed by "test-" followed some name for the test file.
* Test files may appear in either the toplevel package dir (i.e. `lodash/test-lodash.js`), or in the version-specific dir (i.e. `lodash/v4/test-lodash.js`), or in the flow-version-specific dir (i.e. `lodash/v4/flow-v0.22/test-lodash.js`).

### Testing

Given the aforementioned dir structure, there should always be at least one test file for each libdef submission. Tests run as follows:

A test file should exercise the library definition to a suitable degree (when accepting a PR submission, we'll use good judgement here). This includes testing APIs in ways that both expectedly do *not* incur type errors, and ways that do. For type errors that are expected, we'll use a `$ExpectError` suppression comment. For a given test file to pass, it must produce no Flow errors when run with all relevant libdefs.

To run tests: For each `package/version/flow-version` subdirectory, run the specified version of Flow with a temporarily modified version of the .flowconfig which adds all relevant (outer) `test-*` files to the `[includes]` section. If no Flow errors are surfaced, the `package/version/flow-version` is considered passing (and can be published). Obviously we'll script this out and run tests in Travis to ensure that we maintain quality.

**Typically, tests at the package-version level are preferred to flow-version-specific tests.** Flow-version-specific tests are only meant for cases where a new version of Flow adds or fixes some libdef features that weren't available in prior versions.

### Retrieving Library Definitions

We should get the testing and submission infrastructure set up and in place first and foremost (it serves as the foundation of what follows), but once that is in a reasonable state we should be primed to build automated tooling for retrieving library definitions for all stated dependencies of a given project (assuming libdefs for said project/version exist).

Two options: We can either do this with an explicit CLI or we can explore less explicit automated tools.

If we choose to go with an explicit CLI, this means developers would (a) `npm install -g flow-typed` (b) run something like `cd MyPackage && flow-typed install` to analyze + download all available libdefs for their project.

Alternatively (and my preferred solution), we can go with a more automated approach by way of an npm package intended to be installed as a dev-dependency exactly once. This package would have a `postinstall` script that analyzed the depending project's `package.json`, extracts a list of package/version pairs, and automatically installs the corresponding libdefs into the project -- all at `npm install` time.

The reason I prefer the second solution is that it needs to be set up only once and it slipstreams nicely into the existing `npm install` step that people already need to do.

Whichever approach we choose, the retrieval of a libdef for a given Package/Version pair should be done by getting the libdef for the *lowest-compatible-package-version* as stated in the project's `package.json` file. [(See the later sections of this comment for details on why this is important)](https://github.com/splodingsocks/FlowTyped/issues/4#issuecomment-193817961).

If a developer retrieves a libdef using a particular version of Flow (let's say they're using Flow v0.X.Y), and there is both a "flow-all" and a matching "flow-v0.X.Y" version of the requested libdef, the "flow-v0.X.Y" version of the libdef is chosen over the "flow-all" version.

### Publishing Library Definitions

Pull request! We'll make sure we outline all of the requirements for contributing a libdef (tests, naming conventions, etc) -- and tests + automated CI will help us maintain order in that area.

It's important that we have a well-defined (and enforced) structure convention for libdefs in order to support the eventual goal of being able to simplify and automate the process of retrieving libdefs as much as possible.

**Comment by marudor** [comment]
Looks fine to me.
I'm definitively for the 2nd approach of using a postinstall script.
Only question is how should updates be handled? (Updates to libs that is)
If the libs for (as an example) lodash get updated to be more precise/fix some edge case that was not handled before or whatever it will be distributed the next time someone uses npm install.
But a postinstall script definitively sounds better than a new explicit CLI.

The structure itself should give us a lot of flexibility while still easily allowing the most basic approach.

**Comment by magicpat** [comment]
@jeffmo This looks really awesome! 
So for the npm release, are all libdefs of the FlowTyped repository already included?

And for the testing, is it easy to switch between different `flow` versions? Today I realized that it's hard to maintain different projects with different `flow` requirements and a globally installed `flow` binary...

**Comment by jeffmo** [comment]
@marudor 

> Only question is how should updates be handled? (Updates to libs that is)

This is a good question. One thing I left out (mostly to keep discussion a little focused) is that eventually (read: probably not v0?) it would be cool to have a bot running that periodically scans npm for new versions of packages that have a toplevel dir in FlowTyped.

If a new version is discovered *and* Flow is able to reasonably infer an interface definition (i.e. if the package has .js.flow files, this is easy), the bot generates a libdef from that and just checks it in (not sure what to do about tests here -- but given that the libdefs are autogenerated, maybe that's a good enough quality marker).

If a new version is discovered and Flow cannot reasonably infer an interface definition "automatically", then we could add it to a list somewhere or something (especially if it's not covered by an existing libdef version). I would guess that *most* library updates are minor-version updates, so for those it should always be safe to fall back to a libdef for the prior minor version (the libdef might just lack a few things -- but easily fixed by a dev by creating an updated libdef for the new minor). For major versions that can't be auto-inferred, it would be just as if there were no libdef for the package and no libdef would be found.

@magicpat 

> is it easy to switch between different flow versions?

Good question. I think we (the Flow team) have some work to do here, and I know several people on the team are hacking on ways to manage this.

What we do internally at FB is we put a `[version]` entry in all of our .flowconfig files. This does two things: (1) It will cause Flow to error if the .flowconfig version doesn't match the version of Flow you're running and (2) We use a wrapper-script around the globally-installed `flow` CLI that reads this version, finds the real bin for that version on the FS (usually at a known location), and then executes that correct version. I know that @mroch is working on turning this into something more general that we could release to open-source -- so if we can make some headway there, it would be cool to make `flow-bin` do something along these lines.

That said, for the purposes of `FlowTyped`, the installation mechanism would probably just run `flow version` in the context of the local project (or, if available, check the `[version]` of the .flowconfig) to see what Flow version is being used and use that to match against what it queries for in the repo.

**Comment by splodingsocks** [comment]
@jeffmo I think this is a great proposal.

### Re: updating 🎯

I like the second approach a lot. And to handle updating, we could also provide a CLI that, when run, checks for new deps in the project package that don't yet have a flow definition downloaded, and will pull those down. This CLI would be available globally if installed that way, or from within `node_modules/.bin/flow-typed`

### Re: downloading all definitions 📦

If I understood correctly, @magicpat asked if all libdefs would be downloaded to disk when a user runs `npm install flow-typed`. It seems like a good approach would be to have the NPM package only contain the CLI, and have the post-install script be responsible for fetching the appropriate files from the Github Repo directly (or a Couch DB instance, if we got really huge and wanted to be cool like NPM)

### Re: what to do next 🐮

I think we should get a branch made, with the current definitions moved into the suggested file structure, and then we can work on getting the tests running on Travis with different flow versions as appropriate.

Any thoughts on those things?

**Comment by jeffmo** [comment]
Consider me on-board with everything you said!

Having the npm package work dually as a CLI and a downloader seems really simple and clean and probably gives us the best of both approaches

**Comment by splodingsocks** [comment]
Super! Anyone want to take a look at #11 and see if I interpreted the approach correctly?

**Comment by jeffmo** [comment]
PR for the CLI that validates libdefs per the expected structure and runs in Travis: https://github.com/flowtype/flow-typed/pull/14

Still todo: 
* Add a `run-tests` command to the CLI that runs at Travis-time
* Add `/flow-typed` as a default [lib] entry to Flow (if such a dir exists)
* Build the aforementioned postinstall script that fetches appropriately-versioned libdefs from github and puts them into `/flow-typed`

**Comment by ryyppy** [comment]
For the `run-tests` command....
So we need a specific set of flowtype-binary versions to test through the definitions... would it be a good idea to write a script which downloads / builds a specific subset of flowtype binaries (i.e. >= 0.19) and puts it in some `bin` folder inside the flow-types node-module (development only)?

Then we just need a script which takes these specific versions from this local bin... or is it more complicated than I anticipate?

**Comment by jeffmo** [comment]
In fact, I'm about 80% of the way through writing the run-tests script that does just that :)

For each test it basically copies the libdefs+test files into a temporary dir, solves the set of versions of Flow that need to run on that for, grabs those versions of Flow (from the GitHub releases URL), and runs 'flow check' for each of them.

(I'm out of town for the weekend so I likely won't be able to finish and send the PR for review until Monday, but happy to discuss more thoughts here if there are any)

**Comment by ryyppy** [comment]
@jeffmo man, you are a machine! I try to find something to contribute here, so I might as well just write some `libdef` declaration files :-)

**Comment by jeffmo** [comment]
libdef files would be awesome (I wonder which ones are in highest demand and still missing?).

Also if you want to start working on the postinstall downloader script, that would be pretty sweet too

**Comment by jeffmo** [comment]
Just an update: It still needs clean-up, but I now have a branch with a [working test-runner](https://github.com/jeffmo/flow-typed/tree/test_runner).

It's still dev-time messy, so I'll work through cleaning up the code tomorrow to get it ready for a PR.

**Comment by jeffmo** [comment]
Also possibly worth considering an alternate format for specifying the expected compatible Flow version(s). Right now it's a bit awkward to have directories named things like `SomeLib/v1.x/flow-v>=0.20`

**Comment by jeffmo** [comment]
Test runner PR here: https://github.com/flowtype/flow-typed/pull/15

Still haven't gotten to debugging the Travis issues, but worth reviewing the PR in the meantime if anyone has some time to take a look

**Comment by splodingsocks** [comment]
Awesome, @jeffmo! I'll take a look in the morning. Thank you!

**Comment by bsr203** [comment]
Can anyone provide a tldr on how to consume, for example, redux definition today.

Is it like
1. git clone this repo somewhere
2. Add required definition in `[include]` section of `.flowconfig`, or symbolic link to interfaces folder??
It will have a bit awkward path name like `definitions/redux_v3.x.x/flow_>%3Dv0.23.x/redux_v3.x.x.js`

thanks

**Comment by jeffmo** [comment]
@bsr203: Good question. At the moment things are a bit awkward because we haven't shipped the CLI yet (aiming for that next week I hope).

In the meantime, the process is:

1. Create a dir called `flow-typed` in the same directory as your `.flowconfig` file. 
1. Copy the libdef you're interested in from this git repo into that directory.
1. Profit

Flow v0.24 and later will automatically recognize this `flow-typed` directory as an entry in `[libs]` (in other words, no need for config -- just create the directory, put library definitions in it, and Flow will pick them up).

After we ship the CLI, step 2 will be a little less hand-wavy and the workflow will look more like: `flow-typed install myPackage@v0.2.1`

**Comment by valscion** [comment]
> Copy the libdef you're interested in from this git repo into that directory.

I'm sorry, but I don't quite yet understand what files precisely you mean with "Copy the libdef". I've had trouble recognizing classnames declarations, and I'm not sure how my `./flow-typed/` directory should look like.

**Comment by ryyppy** [comment]
Wow, this topic is already reallly really old!
Actually, to answer more appropriately, as @jeffmo already wrote, `flow-typed` is already released as CLI tool (although it's still in beta).

So what you really should do is:
```js
npm install flow-typed -g
cd your-project

flow-typed install classnames@2.0.0 -f 0.25
```

The last command will eventually create your flow-typed directory and put the `classnames` definition in `flow-typed/npm`.

If you have an empty `lib` section in your `.flowconfig`, flow will automatically look into the `flow-typed` dir... if you already defined some lib paths, you have to append `flow-typed/` to this section (I think the auto-assuming is only available for flow >= 0.25).

Whatever you do, don't copy the files manually, that's just a chore :-)

**Comment by ryyppy** [comment]
@splodingsocks @jeffmo ... should we close this issue and maybe change the first README lines with an appropriate show-case example? (thinking about an install command?)

**Comment by valscion** [comment]
> So what you really should do is:
> 
> `npm install flow-typed -g`

Eesh, doesn't seem to be quite easy. Running on OS X El Capitan, npm version 2.15.2 and node v4.4.0, I get this:

```
$ npm install flow-typed -g
-
> nodegit@0.13.2 install /Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit
> node lifecycleScripts/install

[nodegit] Fetching binary from S3.
[nodegit] Failed to install prebuilt binary:
{ [Error: Command failed: /bin/sh -c "/Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit/node_modules/.bin/node-pre-gyp" install --fallback-to-build=false
module.js:327
    throw err;
    ^

Error: Cannot find module 'nopt'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit/node_modules/node-pre-gyp/lib/node-pre-gyp.js:14:12)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
]
  killed: false,
  code: 1,
  signal: null,
  cmd: '/bin/sh -c "/Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit/node_modules/.bin/node-pre-gyp" install --fallback-to-build=false' }
[nodegit] Building manually. (You'll be here a while.)
[nodegit] Regenerating and configuring code
nodegit@0.13.2 /Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/nodegit
├── promisify-node@0.3.0 (nodegit-promise@4.0.0)
└── fs-extra@0.26.7 (path-is-absolute@1.0.0, klaw@1.3.0, graceful-fs@4.1.4, jsonfile@2.3.1, rimraf@2.5.2)

[nodegit] Configuring libssh2.
ERROR - Could not generate native code
[TypeError: _(...).forEach(...).value is not a function]
[nodegit] Everything is ready to go, attempting compilation
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: spawn /Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit/node_modules/.bin/node-gyp ENOENT
    at exports._errnoException (util.js:870:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:178:32)
    at onErrorNT (internal/child_process.js:344:16)
    at nextTickCallbackWith2Args (node.js:442:9)
    at process._tickCallback (node.js:356:17)
npm ERR! Darwin 15.5.0
npm ERR! argv "/Users/vesa/.nvm/versions/node/v4.4.0/bin/node" "/Users/vesa/.nvm/versions/node/v4.4.0/bin/npm" "install" "flow-typed" "-g"
npm ERR! node v4.4.0
npm ERR! npm  v2.15.2
npm ERR! code ELIFECYCLE
```

Do you want me to open a new issue for that?

**Comment by jeffmo** [comment]
> should we close this issue and maybe change the first README lines

Yea, good idea. Will do in just a second...

> Do you want me to open a new issue for that?

Hmm, yes please do. I haven't seen this before so will need some time to dive in

**Comment by ryyppy** [comment]
@valscion Hmm... strange, so I could install it successfully with node version 4.0.0... then I tried it with 4.4.0 and it failed for me as well...
